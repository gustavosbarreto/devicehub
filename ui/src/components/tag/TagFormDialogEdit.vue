<template>
  <fragment>
    <v-list-item-icon class="mr-0">
      <v-icon
        left
        data-test="edit-icon"
        v-text="'mdi-tag'"
      />
    </v-list-item-icon>

    <v-list-item-content>
      <v-list-item-title
        class="text-left"
        data-test="edit-title"
        v-text="'Edit'"
      />
    </v-list-item-content>

    <v-dialog
      v-model="showDialog"
      max-width="400"
      @click:outside="close"
    >
      <v-card data-test="tagForm-card">
        <v-card-title
          class="headline primary"
          data-test="text-title"
          v-text="'Edit tag'"
        />

        <ValidationObserver
          ref="obs"
          v-slot="{ passes }"
        >
          <v-card-text>
            <ValidationProvider
              v-slot="{ errors }"
              ref="providerTag"
              name="Name"
              rules="required|tag|routeIdentifier"
            >
              <v-text-field
                v-model="tagLocal"
                label="Name"
                :error-messages="errors"
                required
                data-test="name-field"
              />
            </ValidationProvider>
          </v-card-text>

          <v-card-actions>
            <v-spacer />

            <v-btn
              text
              data-test="cancel-btn"
              @click="close"
              v-text="'Cancel'"
            />

            <v-btn
              text
              data-test="edit-btn"
              @click="passes(edit)"
              v-text="'Edit'"
            />
          </v-card-actions>
        </ValidationObserver>
      </v-card>
    </v-dialog>
  </fragment>
</template>

<script>

import {
  ValidationObserver,
  ValidationProvider,
} from 'vee-validate';

export default {
  name: 'TagFormDialogComponent',

  components: {
    ValidationProvider,
    ValidationObserver,
  },

  props: {
    uid: {
      type: String,
      default: '',
      required: false,
    },

    tagName: {
      type: String,
      default: '',
      required: false,
    },

    show: {
      type: Boolean,
      required: true,
    },
  },

  data() {
    return {
      dialog: false,
      tagLocal: '',
    };
  },

  computed: {
    showDialog: {
      get() {
        return this.show;
      },

      set(value) {
        this.$emit('update:show', value);
      },
    },
  },

  async updated() {
    await this.setLocalVariable();
  },

  methods: {
    setLocalVariable() {
      this.tagLocal = this.tagName;
    },

    async edit() {
      try {
        await this.$store.dispatch('tags/edit', { oldTag: this.tagName, newTag: this.tagLocal });

        this.update();
        this.$store.dispatch('snackbar/showSnackbarSuccessAction', this.$success.deviceTagEdit);
      } catch {
        this.$store.dispatch('snackbar/showSnackbarErrorAction', this.$errors.snackbar.deviceTagEdit);
      }
    },

    update() {
      this.$emit('update');
      this.close();
    },

    close() {
      this.$emit('update:show', false);
      this.$refs.obs.reset();
    },
  },
};
</script>
