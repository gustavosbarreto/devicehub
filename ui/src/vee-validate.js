import {
  required,
  integer,
  email,
  confirmed,
} from 'vee-validate/dist/rules';
import { extend } from 'vee-validate';
import isValidHostname from 'is-valid-hostname';

import { parsePrivateKey, parseKey } from '@/sshpk';

extend('required', {
  ...required,
  message: 'This field is required',
});

extend('integer', {
  ...integer,
  message: 'This value must be a integer number',
});

extend('email', {
  ...email,
  message: 'This field must be a valid email',
});

extend('noDot', {
  validate: (value) => !/\./.test(value),
  message: 'The name must not contain dots',
});

extend('rfc1123', {
  validate: (value) => isValidHostname(value),
  message: 'You entered an invalid RFC1123 name',
});

extend('routeIdentifier', {
  validate: (value) => !/\/|@|&|:/.test(value),
  message: 'The name must not contain /, @, &, and :.',
});

extend('password', (value) => {
  if (value.length < 5 || value.length > 30) {
    return 'Your password should be 5-30 characters long';
  }
  return true;
});

extend('namespace', (value) => {
  if (value.length < 3 || value.length > 30) {
    return 'Your namespace should be 3-30 characters long';
  }
  return true;
});

extend('device', (value) => {
  if (value.length < 3 || value.length > 30) {
    return 'Your hostname should be 3-30 characters long';
  }
  return true;
});

extend('tag', (value) => {
  if (value.length < 3 || value.length > 255) {
    return 'Your tag should be 3-255 characters long';
  }
  return true;
});

extend('tagRequired', {
  ...required,
  message: 'You must choose at least one tag',
});

extend('tagsLength', {
  validate(value) {
    if (value.length > 3) {
      return 'The maximum capacity has reached';
    }
    return true;
  },
});

extend('comparePasswords', {
  validate(value, { currentPassword }) {
    if (value === currentPassword) {
      return false;
    }
    return true;
  },
  params: ['currentPassword'],
  message: 'The passwords are the same',
});

extend('confirmed', {
  ...confirmed,
  message: 'The passwords do not match',
});

extend('parseKey', {
  validate(value, args) {
    try {
      if (args.typeKey === 'private') {
        parsePrivateKey(value);
      } else {
        parseKey(value);
      }

      return true;
    } catch (err) {
      return false;
    }
  },
  params: ['typeKey'],
  message: 'Not valid key',
});
