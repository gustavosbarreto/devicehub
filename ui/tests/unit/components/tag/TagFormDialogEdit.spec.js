import Vuex from 'vuex';
import { mount, createLocalVue } from '@vue/test-utils';
import { ValidationProvider, ValidationObserver } from 'vee-validate';
import flushPromises from 'flush-promises';
import Vuetify from 'vuetify';
import TagFormDialogEdit from '@/components/tag/TagFormDialogEdit';
import '@/vee-validate';

describe('TagFormDialogEdit', () => {
  const localVue = createLocalVue();
  const vuetify = new Vuetify();
  localVue.use(Vuex);
  localVue.component('ValidationProvider', ValidationProvider);
  localVue.component('ValidationObserver', ValidationObserver);

  document.body.setAttribute('data-app', true);

  let wrapper;

  // vee-validate variables bellow
  const invalidName = ['xxx/', 'xxx@', 'xxx&', 'xxx:'];
  const invalidMinAndMaxCharacters = [
    'x', 'xx',
  ];

  const tests = [
    {
      description: 'Dialog closed',
      props: {
        uid: 'b7:25:f8',
        tagName: 'ShellHub',
        show: false,
      },
      data: {
        dialog: false,
        tagLocal: '',
      },
      template: {
        'edit-icon': true,
        'edit-title': true,
        'edit-itemtagForm-card': false,
      },
      templateText: {
        'edit-title': 'Edit',
      },
    },
    {
      description: 'Dialog Opened',
      props: {
        uid: 'b7:25:f8',
        tagName: 'ShellHub',
        show: true,
      },
      data: {
        dialog: false,
        tagLocal: '',
      },
      template: {
        'edit-icon': true,
        'edit-title': true,
        'tagForm-card': true,
        'text-title': true,
        'name-field': true,
        'cancel-btn': true,
        'edit-btn': true,
      },
      templateText: {
        'edit-title': 'Edit',
        'text-title': 'Edit tag',
        'cancel-btn': 'Cancel',
        'edit-btn': 'Edit',
      },
    },
  ];

  const storeVuex = () => new Vuex.Store({
    namespaced: true,
    state: { },
    getters: { },
    actions: {
      'tags/edit': () => {},
      'snackbar/showSnackbarSuccessAction': () => {},
      'snackbar/showSnackbarErrorAction': () => {},
    },
  });

  tests.forEach((test) => {
    describe(`${test.description}`, () => {
      beforeEach(() => {
        wrapper = mount(TagFormDialogEdit, {
          store: storeVuex(),
          localVue,
          stubs: ['fragment'],
          propsData: {
            uid: test.props.uid,
            tagName: test.props.tagName,
            show: test.props.show,
          },
          vuetify,
        });
      });

      ///////
      // Component Rendering
      //////

      it('Is a Vue instance', () => {
        expect(wrapper).toBeTruthy();
      });
      it('Renders the component', () => {
        expect(wrapper.html()).toMatchSnapshot();
      });

      ///////
      // Data checking
      //////

      it('Receive data in props', () => {
        Object.keys(test.props).forEach((item) => {
          expect(wrapper.vm[item]).toEqual(test.props[item]);
        });
      });
      it('Compare data with default value', () => {
        Object.keys(test.data).forEach((item) => {
          expect(wrapper.vm[item]).toEqual(test.data[item]);
        });
      });

      //////
      // HTML validation
      //////

      it('Renders the template with data', () => {
        Object.keys(test.template).forEach((item) => {
          expect(wrapper.find(`[data-test="${item}"]`).exists()).toBe(test.template[item]);
        });
      });
      it('Renders template with expected text', () => {
        Object.keys(test.templateText).forEach((item) => {
          expect(wrapper.find(`[data-test="${item}"]`).text()).toContain(test.templateText[item]);
        });
      });

      if (test.props.show) {
        //////
        // In this case, the route identifier are validated.
        //////

        it('Show validation messages', async (done) => {
          invalidName.forEach(async (name) => {
            wrapper.setData({ tagLocal: name });
            await flushPromises();

            const validator = wrapper.vm.$refs.providerTag;

            await validator.validate();
            expect(validator.errors[0]).toBe('The name must not contain /, @, &, and :.');

            await flushPromises();
            done();
          });
        });

        //////
        // In this case, the min and max characters are validated.
        //////

        it('Show validation messages', async (done) => {
          const invalidMaxCharacter = 'x'.repeat(256);

          invalidMinAndMaxCharacters.push(invalidMaxCharacter);
          invalidMinAndMaxCharacters.push(invalidMaxCharacter.concat('x'));

          invalidMinAndMaxCharacters.forEach(async (character) => {
            wrapper.setData({ tagLocal: character });
            await flushPromises();

            const validator = wrapper.vm.$refs.providerTag;

            await validator.validate();
            expect(validator.errors[0]).toBe('Your tag should be 3-255 characters long');

            await flushPromises();
            done();
          });
        });
      }
    });
  });
});
