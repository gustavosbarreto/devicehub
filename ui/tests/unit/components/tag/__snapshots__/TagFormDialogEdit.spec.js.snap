// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagFormDialogEdit Dialog Opened Renders the component 1`] = `
<fragment-stub>
  <div class="v-list-item__icon mr-0"><i aria-hidden="true" data-test="edit-icon" class="v-icon notranslate v-icon--left mdi mdi-tag theme--light"></i></div>
  <div class="v-list-item__content">
    <div data-test="edit-title" class="v-list-item__title text-left">Edit</div>
  </div>
  <div class="v-dialog__container"></div>
</fragment-stub>
`;

exports[`TagFormDialogEdit Dialog closed Renders the component 1`] = `
<fragment-stub>
  <div class="v-list-item__icon mr-0"><i aria-hidden="true" data-test="edit-icon" class="v-icon notranslate v-icon--left mdi mdi-tag theme--light"></i></div>
  <div class="v-list-item__content">
    <div data-test="edit-title" class="v-list-item__title text-left">Edit</div>
  </div>
  <div class="v-dialog__container">
    <!---->
  </div>
</fragment-stub>
`;
